### PR TITLE
Feat/create and update tour

### PR DESCRIPTION
**Pull request by QuocThangPhu**
create function CreateAndUpdateTourController

### CODE REVIEW

File: src/Controller/API/TourController.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    #[IsGranted('ROLE_ADMIN')]
+    public function addTour(
+        Request $request,
+        TourRequest $tourRequest,
+        TourValidator $tourValidator,
+        TourService $tourService,
+        TourTransformer $tourTransformer,
+    ): JsonResponse
+    {
+        $dataRequest = json_decode($request->getContent(), true);
```
Comment (nhivo1203) [inline src/Controller/API/TourController.php:33]: i think use $requestData = $request->toArray() is better

File: src/Request/TourRequest.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    #[Assert\Type('array')]
+    #[Assert\NotBlank]
+    #[Assert\NotNull]
+    private $tourPlan;
+
+    #[Assert\Type('array')]
+    #[Assert\NotBlank]
+    #[Assert\NotNull]
+    private $tourImage;
```
Comment (nhivo1203) [inline src/Request/TourRequest.php:52]: should refactor tourImages, because it's a array

File: .gitignore [hunk 3]
```diff
@@ -22,8 +22,3 @@
 /symfony.lock
 /phpunit.xml.dist
 /composer.lock
-
-###> symfony/phpunit-bridge ###
-.phpunit.result.cache
```
Comment (bangdinhnfq) [inline .gitignore:27]: Why remove?

File: src/Service/TourService.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            $tour->addService($service);
+        }
+        return $tour;
+    }
+
+    private function addTourPlan(TourRequest $tourRequest, Tour $tour): Tour
+    {
+        foreach ($tourRequest->getTourPlan() as $tourPlanRequest) {
+            $destination = $this->destinationRepository->find($tourPlanRequest['destination']);
+            if (is_object($destination)) {
```
Comment (bangdinhnfq) [inline src/Service/TourService.php:69]: Must use early return

File: src/Service/TourService.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                $this->tourPlanRepository->add($tourPlan);
+            }
+        }
+        return $tour;
+    }
+
+    private function addTourImage(TourRequest $tourRequest, Tour $tour): Tour
+    {
+        foreach ($tourRequest->getTourImage() as $tourImageRequest) {
+            $image = $this->imageRepository->find($tourImageRequest['id']);
```
Comment (bangdinhnfq) [inline src/Service/TourService.php:85]: How about the empty case?

File: src/Transformer/TourTransformer.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace App\Transformer;
+
+use App\Entity\Tour;
+
+class TourTransformer extends BaseTransformer
+{
+    private const PARAMS = ['id', 'title', 'duration', 'maxPeople', 'minAge', 'overView', 'price'];
+
+    public function fromArray(Tour $tour): array
```
Comment (bangdinhnfq) [inline src/Transformer/TourTransformer.php:11]: It should be `toArray` ?

File: .gitignore (hunk 3 above)
Comment (QuocThangPhu) [reply .gitignore:27]: because it's was exist in .gitignore

File: src/Request/ListTourRequest.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * @param int $guests
+     */
+    public function setGuests(int $guests): void
+    {
+        $this->guests = $guests;
+    }
+}
```
Comment (khajackie2206) [inline src/Request/ListTourRequest.php:117]: EOL

File: src/Mapper/TourUpdateRequestToTour.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            ->setDuration($tourUpdateRequest->getDuration() ?? $tour->getDuration())
+            ->setMaxPeople($tourUpdateRequest->getMaxPeople() ?? $tour->getMaxPeople())
+            ->setMinAge($tourUpdateRequest->getMinAge() ?? $tour->getMinAge())
+            ->setOverView($tourUpdateRequest->getOverView() ?? $tour->getOverView())
+            ->setPrice($tourUpdateRequest->getPrice() ?? $tour->getPrice())
+            ->setUpdatedAt(new \DateTimeImmutable())
+            ->setCreatedUser($currentUser);
+        return $tour;
+    }
+}
```
Comment (khajackie2206) [inline src/Mapper/TourUpdateRequestToTour.php:36]: hard to understand your naming

File: src/Mapper/TourRequestToTour.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        $tour->setTitle($tourRequest->getTitle())
+            ->setDuration($tourRequest->getDuration())
+            ->setMaxPeople($tourRequest->getMaxPeople())
+            ->setMinAge($tourRequest->getMinAge())
+            ->setOverView($tourRequest->getOverView())
+            ->setPrice($tourRequest->getPrice())
+            ->setCreatedUser($currentUser);
+        return $tour;
+    }
+}
```
Comment (khajackie2206) [inline src/Mapper/TourRequestToTour.php:37]: It's very hard to understand to your naming

**Review by khajackie2206 — APPROVED** [review]

**Pull request merged**